pkg/gatewayserver/io/io.go

```diff
@@ -342,9 +337,19 @@ func (c *Connection) ScheduleDown(path *ttnpb.DownlinkPath, msg *ttnpb.DownlinkM
 		switch request.Class {
 		case ttnpb.CLASS_A:
 			f = c.scheduler.ScheduleAt
-			settings.Timestamp = uplinkTimestamp + uint32(rxDelay/time.Microsecond)
+
+			rx1Delay := time.Duration(request.Rx1Delay) * time.Second
+			if rx1Delay == 0 {
+				rx1Delay = time.Second // RX_DELAY_0 is valid, and 1 second.
+			}
+			settings.Timestamp = uplinkTimestamp + uint32((rx1Delay+rx.delay)/time.Microsecond)
 		case ttnpb.CLASS_B:
-			f = c.scheduler.ScheduleAnytime
+			if request.AbsoluteTime == nil {
+				return 0, errNotAllowed
+			}
+			f = c.scheduler.ScheduleAt
+			abs := *request.AbsoluteTime
+			settings.Time = &abs
```

Comment:
The original idea was that `ScheduleAnytime` would find a suitable ping slot based on the `TxRequest`. We could decorate `TxRequest` further with class B things, like future ping slots, so that GS can choose, taking conflicts and duty-cycle constraints into account. Or do you want NS to know exactly which ping slot is used? In that case, this is correct.

Not sure what `abs` is for though

Reply:
> Not sure what `abs` is for though

Me neither, that's just copy-paste from class C.

> The original idea was that `ScheduleAnytime` would find a suitable ping slot based on the `TxRequest`

That will not work and it's out of scope of GS. In order to compute the ping slot time, knowledge of device's MAC state is required - it's NS's job to do that.
I don think that GS should not use it's own clock here, but use the gateway's GPS-synced time instead when scheduling.

Reply:
> That will not work and it's out of scope of GS. In order to compute the ping slot time, knowledge of device's MAC state is required - it's NS's job to do that.

OK

> I think that GS should not use it's own clock here, but use the gateway's GPS-synced time instead when scheduling.

FTR, there is no GS clock; GS syncs `tmst` with `tmms` and always schedules with `tmst` instead of `tmms`.